src/local_executor.rs

```diff
@@ -0,0 +1,70 @@
+//! Execution of futures on a single thread
+//!
+//! This module has no special handling of any blocking operations other than
+//! futures-aware inter-thread communications, and should therefore probably not
+//! be used to manage IO.
+
+use std::sync::{Arc, Mutex, mpsc};
+use std::collections::HashMap;
+use std::collections::hash_map;
+use std::boxed::Box;
+
+use {Future, Async};
+
+use executor::{self, Spawn};
+
+/// Main loop object
+pub struct Core {
+    unpark_send: mpsc::Sender<u64>,
+    unpark: mpsc::Receiver<u64>,
+    live: HashMap<u64, Spawn<Box<Future<Item=(), Error=()>>>>,
+    next_id: u64,
+}
+
+impl Core {
+    /// Create a new `Core`.
+    pub fn new() -> Self {
+        let (send, recv) = mpsc::channel();
+        Core {
+            unpark_send: send,
+            unpark: recv,
+            live: HashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    /// Spawn a future to be executed by a future call to `run`.
+    pub fn spawn<F>(&mut self, f: F)
+        where F: Future<Item=(), Error=()> + 'static
+    {
+        self.live.insert(self.next_id, executor::spawn(Box::new(f)));
+        self.unpark_send.send(self.next_id).unwrap();
+        self.next_id += 1;
+    }
+
+    /// Run the loop until all futures previously passed to `spawn` complete.
+    pub fn run(&mut self) {
```

Comment:
This file looks pretty god! Should definitely work for running tests. If you feel the need to you can also add `F` as a type parameter to this `run` method which basically runs everything until that future completes.